### DISCUSSION TITLE
Freedom能否不解析域名直连IP？

### DISCUSSION THREAD

**Discussion by Smoovsky** [XTLS/Xray-core#2163] [Q&A]
目前看来行为和文档描述的是一致的：
```
"AsIs": Freedom 通过系统 DNS 服务器解析获取 IP, 向此域名发出连接.
```
但是如果我希望即便sniffing出来域名但是走Freedom outbound的时候仍然直连IP，请问有什么办法能办到吗？

之所以有这个需求是因为有一些在CDN后面的网站我希望保持延迟尽可能低，所以本地写了个脚本用来检测哪个CDN服务器的延迟最低，然后直连CDN服务器的IP并且手动指定sni（例如直接修改hosts文件）。但是因为xray/v2ray的freedom outbound的这个行为，只要tls连接里嗅探出来了sni，目标IP就会被重置成本地dns的解析结果。但是sni又是不能被省略的，否则CDN反代就不能成功。

**Comment by Smallthing** [comment]
你弄的这么复杂都不如本地弄个优选dns缓存

**Reply by Smoovsky** [reply]
源站在境外，cdn在境内，所以比起本地到cdn的延迟，cdn到源站的影响要大得多。几个主流的公共dns解析出来的结果大多也不甚理想。
所以就只能把cdn所有的服务器ip都dig出来然后用定期脚本优选了，优选出来的结果会写入dnsmasq的hosts文件供局域网其他机器使用。这部分没什么问题，可以按照预期的工作。
但是目前因为xray freedom outbound这个行为导致不能直接在旁路由上跑这个脚本。

**Comment by chise0713** [comment]
> 但是如果我希望即便sniffing出来域名但是走Freedom outbound的时候仍然直连IP，请问有什么办法能办到吗？

使用 `routeOnly` 可以达到你想要的效果
```jsonc
{
  "enabled": true,
  "destOverride": ["http", "tls", "quic"],
  "metadataOnly": false, 
  "domainsExcluded": [],
  "routeOnly": true  // 打开
}
```

**Comment by us254** [comment]
you can try using the "UseIP" mode in Xray's freedom outbound configuration. By specifying the "UseIP" mode and setting the sendThrough field in the outbound connection configuration, Freedom will automatically determine the required IP type (IPv4 or IPv6) based on the value of sendThrough.


```json
{
  "outbound": {
    "protocol": "freedom",
    "settings": {
      "domainStrategy": "UseIP"
    },
    "tag": "direct-out"
  }
}
```

By using the "UseIP" mode, Xray will resolve the domain name using the built-in DNS server and connect directly to the obtained IP. This should allow you to maintain a direct connection to the IP of the CDN server, even if the domain name is sniffed.
